### PR TITLE
Nightly builds

### PR DESCRIPTION
**Pull request by strickvl**
These changes set up the following:
- a scheduled run at midnight every night
- this first runs the unit tests on Ubuntu to prevent pushing something truly broken
- then it pushes the package as `zenml-nightly` with the version e.g. 0.46.0-2023-11-07
- it then pushes all the docker images (server and Python 3.10) with the `nightly` tag

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: .github/workflows/publish_to_pypi_nightly.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+              # Update the version to the nightly version
+              poetry version $NIGHTLY_VERSION
+
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.8'
+
+      - name: Install Poetry
```
Comment (schustmi) [inline .github/workflows/publish_to_pypi_nightly.yml:48]: How can the previous steps even work if poetry is installed afterwards? 🤔

File: .github/workflows/publish_docker_image_nightly.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+          service_account_email: ${{ secrets.GCP_CLOUDBUILD_EMAIL }}
+          service_account_key: ${{ secrets.GCP_CLOUDBUILD_KEY }}
+          project_id: ${{ secrets.GCP_CLOUDBUILD_PROJECT }}
+
+      # Cloudbuild
+      - name: Build docker images
+        run: |-
+          gcloud builds submit \
+            --quiet \
+            --config=release-cloudbuild-nightly.yaml \
```
Comment (schustmi) [inline .github/workflows/publish_docker_image_nightly.yml:35]: Can we reuse the other workflow and make the config configurable? Or are there more changes in this file which complicate things?

**Review by avishniakov — APPROVED** [review]
Looks good, I'll trust it's working 👍🏼

File: release-cloudbuild-nightly.yaml [hunk 3]
```diff
@@ -0,0 +1,138 @@
+steps:
```
Comment (schustmi) [inline release-cloudbuild-nightly.yaml:1]: I'm wondering in general whether we need to release/build Docker images for all python versions for these nightly builds?

Is the idea behind it to make sure our release workflows run without issues, or are we expecting people to actively use these releases in environments in which they're restricted to a specific python version?

**Review by schustmi — CHANGES_REQUESTED** [review]

File: release-cloudbuild-nightly.yaml (hunk 3 above)
Comment (strickvl) [reply release-cloudbuild-nightly.yaml:1]: The main use case is internal, tbh, for demos. But it is also part of our reworking of the test process in general. I'd be fine only pushing one tagged image for now @ Python 3.10 or something. We can always add more if people ask for them.

File: .github/workflows/publish_docker_image_nightly.yml (hunk 2 above)
Comment (strickvl) [reply .github/workflows/publish_docker_image_nightly.yml]: You're right. I've refactored it away to take an argument that specifies the config file. That was the only thing that distinguishes the two workflows.

File: .github/workflows/nightly_build.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        runs-on: ubuntu-latest
+        needs: publish-python-package
+        steps:
+          - name: Sleep for 4 minutes
+            run: sleep 240
+            shell: bash
+
+    publish-docker-image:
+        needs: wait-for-package-release
+        uses: ./.github/workflows/publish_docker_image_nightly.yml
```
Comment (schustmi) [inline .github/workflows/nightly_build.yml:33]: This should not have the nightly suffix I think?

Comment (schustmi) [reply .github/workflows/nightly_build.yml:33]: I think we also need to pass the normal release config in the existing release workflow?

**Review by schustmi — CHANGES_REQUESTED** [review]

File: .github/workflows/nightly_build.yml (hunk 4 above)
Comment (strickvl) [reply .github/workflows/nightly_build.yml:33]: No need for the normal release config to be passed in for the existing release workflow as it's specified explicitly as the default in `publish_docker_image.yml`. But the nightly suffix can be deleted, yes.

**Review by stefannica — APPROVED** [review]
LGTM. We might need to include other python versions when/if people start asking for them.

**Review by schustmi — APPROVED** [review]

**Pull request merged**